main.go

```diff
@@ -0,0 +1,170 @@
+package main
+
+import (
+	"bufio"
+	"encoding/json"
+	"fmt"
+	"log"
+	"os"
+	"sort"
+)
+
+type Entry struct {
+	ID          uint32
+	FirstName   string
+	LastName    string
+	PhoneNumber string
+}
+
+func mainMenu() (choice int) {
+	fmt.Println(`Choose your action:
+1) List all entries.
+2) Add new entry.
+3) Remove an entry by ID.
+4) Quit`)
+	fmt.Scanln(&choice)
+	return choice
+}
+
+func printHeader() {
+	fmt.Printf("%10s %20s %20s %20s\n", "ID", "Last Name", "First Name", "Phone Number")
+	fmt.Println("-------------------------------------------------------------------------")
+}
+
+func listEntries() {
+	var entries []Entry
+	f, err := os.Open("phonebook.json")
+	if err != nil {
+		log.Fatal(err)
+	}
+	defer f.Close()
+	if err := json.NewDecoder(bufio.NewReader(f)).Decode(&entries); err != nil {
+		log.Fatal(err)
+	}
+	sort.Slice(entries, func(i, j int) bool { return entries[i].LastName < entries[j].LastName })
+
+	printHeader()
+	for i, e := range entries {
+		fmt.Printf("%10d %20s %20s %20s\n", e.ID, e.LastName, e.FirstName, e.PhoneNumber)
+		if (i+1)%20 == 0 && i < len(entries)-1 { // Avoid the prompt for the last row.
+			fmt.Print("Press <ENTER> to continue...")
+			fmt.Scanln()
+			printHeader()
+		}
+	}
+}
+
+func newEntry() {
+	var entries []Entry
+	f, err := os.Open("phonebook.json")
```

Comment:
It is always a good idea to have reading and writing in dedicated functions. `defer` postpones this call till the end of the function. It means that you attempt to create and write to a file which is already opened for reading. Some operating systems would fail writing to a file while it's opened for reading.

In addition to this, having it factored out just reads better:

```go
func loadPhonebook() ([]Entry, error) { ... }
func savePhonebook([]Entry) error { ... }
func newEntry() {
  entries, err := loadPhonebook()
  if err != nil { log.Fatal(err) }

  e := Entry{ID: maxID(entries)+1}
  fmt.Print("Enter last name: "); fmt.Scanln(&e.LastName)
  fmt.Print("Enter first name: "); fmt.Scanln(&e.FirstName)
  fmt.Print("Enter phone number: "); fmt.Scanln(&e.PhoneNumber)
  entries = append(entries, e)

  if err := savePhonebook(entries); err != nil { log.Fatal(err) }
}
```

see, how it reads now? short & clear